dask/layers.py

```diff
@@ -1239,3 +1257,309 @@ def __repr__(self):
         )
+
+
+class DataFrameTreeReduction(DataFrameLayer):
+    """DataFrame Tree-Reduction Layer
+
+    Parameters
+    ----------
+    name : str
+        Name to use for the constructed layer.
+    name_input : str
+        Name of the input layer that is being reduced.
+    npartitions_input : str
+        Number of partitions in the input layer.
+    concat_func : callable
+        Function used by each tree node to reduce a list of inputs
+        into a single output value. This function must accept only
+        a list as its first positional argument.
+    tree_node_func : callable
+        Function used on the output of ``concat_func`` in each tree
+        node. This function must accept the output of ``concat_func``
+        as its first positional argument.
+    finalize_func : callable, optional
+        Function used in place of ``tree_node_func`` on the final tree
+        node(s) to produce the final output for each split. By default,
+        ``tree_node_func`` will be used.
+    split_every : int, optional
+        This argument specifies the maximum number of input nodes
+        to be handled by any one task in the tree. Defaults to 32.
+    split_out : int, optional
+        This argument specifies the number of output nodes in the
+        reduction tree. If ``split_out`` is set to an integer >=1, the
+        input tasks must contain data that can be indexed by a ``getitem``
+        operation with a key in the range ``[0, split_out)``.
+    output_partitions : list, optional
+        List of required output partitions. This parameter is used
+        internally by Dask for high-level culling.
+    tree_node_name : str, optional
+        Name to use for intermediate tree-node tasks.
+    """
+
+    name: str
+    name_input: str
+    npartitions_input: str
+    concat_func: Callable
+    tree_node_func: Callable
+    finalize_func: Optional[Callable]
+    split_every: Optional[int]
+    split_out: Optional[int]
+    output_partitions: Optional[List[int]]
+    tree_node_name: Optional[str]
+    widths: List[int]
+    height: int
+
+    def __init__(
+        self,
+        name: str,
+        name_input: str,
+        npartitions_input: str,
+        concat_func: Callable,
+        tree_node_func: Callable,
+        finalize_func: Optional[Callable] = None,
+        split_every: Optional[int] = 32,
+        split_out: Optional[int] = None,
+        output_partitions: Optional[List[int]] = None,
+        tree_node_name: Optional[str] = None,
+        annotations: Mapping[str, Any] = None,
+    ):
+        super().__init__(annotations=annotations)
+        self.name = name
+        self.name_input = name_input
+        self.npartitions_input = npartitions_input
+        self.concat_func = concat_func
+        self.tree_node_func = tree_node_func
+        self.finalize_func = finalize_func
+        self.split_every = split_every
+        self.split_out = split_out
+        self.output_partitions = output_partitions
```

Comment:
```suggestion
        self.output_partitions = output_partitions or list(range(self.split_out or 1))
```

Reply:
`output_partitions` can technically be an empty list if everything is culled.  We set the default value below if `self.output_partitions is None`

Reply:
Good point. I just suggested moving this here because with https://github.com/dask/dask/pull/8468#discussion_r789083728, this line would now be a type error.

Reply:
Yeah - good call. Hopefully that error is addressed now